packages/database-types/index.d.ts

```diff
@@ -32,9 +32,16 @@ export interface DataSnapshot {
   val(): any;
 }
 
-export class FirebaseDatabase {
-  private constructor();
+export interface Database {
+  app: FirebaseApp;
+  goOffline(): void;
+  goOnline(): void;
+  ref(path?: string | Reference): Reference;
+  refFromURL(url: string): Reference;
+}
 
+export class FirebaseDatabase implements Database {
```

Comment:
Since this is a .d.ts file, isn't "class" already considered an interface? Could we just drop `private constructor`?

Reply:
IIUC, we use it to type the actual class that we export and we don't want people to instantiate it, so `private constructor` is important. Please correct me if i'm wrong.

Reply:
You can use d.ts file to annotate a JS library. It tells Typescript that the JS library exports a class that people can import and use. People can actually do this and Typescript allows it:

```
import {FirebaseDatabase} from '@firebase/database-types';

new FirebaseDatabase(); // if not having the private constructor
```

For this reason, I consider it's a bad practice to use class instead of interface in our types packages, if you are just going for the instance type.

Reply:
If I understand this correctly, it seems like the following two are equivalent:

```
export class Foo {
  private constructor();
  foo: string;
}

export interface Foo {
 foo: string;
}
```

If so, can we drop the existing class declaration? If not, this LGTM.

Reply:
They are the same for the instance type, but we need the constructor type which requires to define it as a class. See https://github.com/firebase/firebase-js-sdk/blob/master/packages/database/index.ts#L97

I believe it is to allow people to do instanceof, but disallow newing an object.

Reply:
Yeah, the `instanceof` check was the only thing that I could think of, but it seemed a bit unlikely. Better not to break anyone though.